packages/engine/paima-funnel/src/funnels/avail/utils.ts

```diff
@@ -0,0 +1,171 @@
+import type { ApiPromise } from 'avail-js-sdk';
+import type { Header as PolkadotHeader } from '@polkadot/types/interfaces/types';
+import type { SubmittedData } from '@paima/sm';
+import { base64Decode } from '@polkadot/util-crypto';
+import { BaseFunnelSharedApi } from '../BaseFunnel.js';
+import { createApi } from './createApi.js';
+
+export const GET_DATA_TIMEOUT = 10000;
+
+export type Header = PolkadotHeader;
+
+export function getSlotFromHeader(header: Header, api: ApiPromise): number {
+  const preRuntime = header.digest.logs.find(log => log.isPreRuntime)!.asPreRuntime;
+
+  const rawBabeDigest = api.createType('RawBabePreDigest', preRuntime[1]);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const babeDigest = rawBabeDigest.toPrimitive() as unknown as any;
+
+  // the object is an enumeration, but all the variants have a slotNumber field
+  const slot = babeDigest[Object.getOwnPropertyNames(babeDigest)[0]].slotNumber;
+  return slot;
+}
+
+export async function getLatestBlockNumber(api: ApiPromise): Promise<number> {
+  let highHash = await api.rpc.chain.getFinalizedHead();
+  let high = (await api.rpc.chain.getHeader(highHash)).number.toNumber();
+  return high;
+}
+
+export async function getTimestampForBlockAt(api: ApiPromise, mid: number): Promise<number> {
+  const hash = await api.rpc.chain.getBlockHash(mid);
+  // FIXME: why is the conversion needed?
+  const header = (await api.rpc.chain.getHeader(hash)) as unknown as Header;
+
+  const slot = getSlotFromHeader(header, api);
+  return slotToTimestamp(slot, api);
+}
+
+export function slotToTimestamp(slot: number, api: ApiPromise): number {
+  // this is how it's computed by the pallet
+  // https://paritytech.github.io/polkadot-sdk/master/src/pallet_babe/lib.rs.html#533
```

Comment:
note: here and below, these are not permalinks, so this will stop working if they ever modify these files. Probably it's better to use Github permalinks instead